src/inmanta/server/services/orchestrationservice.py

```diff
@@ -384,6 +384,39 @@ async def _purge_versions(self) -> None:
                 for v in delete_list:
                     await version_dict[v].delete_cascade()
 
+        # Cleanup old agents from agent table in db
+        async with data.Agent.get_connection() as con:
+            await con.execute(
+                """
+-- Delete all records from the agent table where the (environment, name) pair
+-- is not in the set of (environment, name) pairs returned by the subquery.
+DELETE FROM public.agent
```

Comment:
I did not. Could you repeat to me how I can check this? I'll add it to the dev docs so I won't forget it next time

Reply:
You can restore a database dump from the tests folder on your local PostgreSQL installation and run the query against that database but with the EXPLAIN ANALYZE prefix. That way postgresql will report which indexes it uses and what the estimated cost of each query is.

Reply:
thanks! this is the result:
```
  Delete on agent  (cost=20.00..21.03 rows=1 width=6) (actual time=0.066..0.068 rows=0 loops=1)
   ->  Seq Scan on agent  (cost=20.00..21.03 rows=1 width=6) (actual time=0.065..0.067 rows=0 loops=1)
         Filter: (NOT (hashed SubPlan 1))
         Rows Removed by Filter: 2
         SubPlan 1
           ->  HashAggregate  (cost=17.50..19.50 rows=200 width=48) (actual time=0.044..0.045 rows=2 loops=1)
                 Group Key: e.id, map.key
                 Batches: 1  Memory Usage: 40kB
                 ->  HashAggregate  (cost=12.35..14.41 rows=206 width=48) (actual time=0.042..0.044 rows=2 loops=1)
                       Group Key: e.id, map.key
                       Batches: 1  Memory Usage: 40kB
                       ->  Append  (cost=0.01..11.32 rows=206 width=48) (actual time=0.012..0.037 rows=6 loops=1)
                             ->  Nested Loop  (cost=0.01..5.02 rows=200 width=48) (actual time=0.012..0.013 rows=2 loops=1)
                                   ->  Seq Scan on environment e  (cost=0.00..1.02 rows=2 width=48) (actual time=0.002..0.002 rows=2 loops=1)
                                   ->  Function Scan on jsonb_each map  (cost=0.01..1.00 rows=100 width=32) (actual time=0.004..0.004 rows=1 loops=2)
                             ->  Subquery Scan on "*SELECT* 2"  (cost=0.00..1.04 rows=2 width=48) (actual time=0.001..0.002 rows=2 loops=1)
                                   ->  Seq Scan on agent agent_1  (cost=0.00..1.02 rows=2 width=48) (actual time=0.001..0.001 rows=2 loops=1)
                                         Filter: (id_primary IS NOT NULL)
                             ->  Subquery Scan on "*SELECT* 3"  (cost=2.15..2.22 rows=4 width=48) (actual time=0.019..0.021 rows=2 loops=1)
                                   ->  Unique  (cost=2.15..2.18 rows=4 width=48) (actual time=0.019..0.020 rows=2 loops=1)
                                         ->  Sort  (cost=2.15..2.16 rows=4 width=48) (actual time=0.019..0.020 rows=4 loops=1)
                                               Sort Key: r.environment, r.agent
                                               Sort Method: quicksort  Memory: 25kB
                                               ->  Hash Join  (cost=1.05..2.11 rows=4 width=48) (actual time=0.009..0.010 rows=4 loops=1)
                                                     Hash Cond: ((r.environment = c.environment) AND (r.model = c.version))
                                                     ->  Seq Scan on resource r  (cost=0.00..1.04 rows=4 width=52) (actual time=0.001..0.001 rows=4 loops=1)
                                                     ->  Hash  (cost=1.02..1.02 rows=2 width=20) (actual time=0.004..0.004 rows=2 loops=1)
                                                           Buckets: 1024  Batches: 1  Memory Usage: 9kB
                                                           ->  Seq Scan on configurationmodel c  (cost=0.00..1.02 rows=2 width=20) (actual time=0.002..0.002 rows=2 loops=1)
 Planning Time: 0.261 ms
 Execution Time: 0.133 ms
(31 rows)
```
So indeed the third select seems to cost more and the NOT has some cost to, but do I understand correctly that the Union call is the one that cost the most? Also I'm not able to tell how "bad" it is

Reply:
Indeed. It's hard to tell from this how bad it is. Especially because the data set is rather small. I did the same on a dump of the database from mgmt and then we get the following:

```
 Seq Scan on agent  (cost=6618.02..6623.90 rows=96 width=69) (actual time=61.036..61.064 rows=46 loops=1)
   Filter: (NOT (hashed SubPlan 1))
   Rows Removed by Filter: 146
   SubPlan 1
     ->  Unique  (cost=6610.84..6617.52 rows=200 width=48) (actual time=60.940..60.973 rows=193 loops=1)
           ->  Sort  (cost=6610.84..6613.07 rows=890 width=48) (actual time=60.939..60.950 rows=380 loops=1)
                 Sort Key: e.id, map.key
                 Sort Method: quicksort  Memory: 58kB
                 ->  Append  (cost=0.01..6558.34 rows=890 width=48) (actual time=0.048..60.651 rows=380 loops=1)
                       ->  Nested Loop  (cost=0.01..13.06 rows=600 width=48) (actual time=0.047..0.194 rows=186 loops=1)
                             ->  Seq Scan on environment e  (cost=0.00..1.06 rows=6 width=48) (actual time=0.008..0.009 rows=6 loops=1)
                             ->  Function Scan on jsonb_each map  (cost=0.01..1.00 rows=100 width=32) (actual time=0.026..0.027 rows=31 loops=6)
                       ->  Subquery Scan on "*SELECT* 2"  (cost=0.00..6.35 rows=143 width=48) (actual time=0.003..0.022 rows=143 loops=1)
                             ->  Seq Scan on agent agent_1  (cost=0.00..4.92 rows=143 width=43) (actual time=0.002..0.014 rows=143 loops=1)
                                   Filter: (id_primary IS NOT NULL)
                                   Rows Removed by Filter: 49
                       ->  Subquery Scan on "*SELECT* 3"  (cost=6525.54..6528.48 rows=147 width=48) (actual time=60.413..60.419 rows=51 loops=1)
                             ->  HashAggregate  (cost=6525.54..6527.01 rows=147 width=44) (actual time=60.412..60.416 rows=51 loops=1)
                                   Group Key: r.environment, r.agent
                                   Batches: 1  Memory Usage: 40kB
                                   ->  Nested Loop  (cost=0.42..4722.16 rows=360675 width=44) (actual time=0.019..28.297 rows=360622 loops=1)
                                         ->  Seq Scan on configurationmodel c  (cost=0.00..13.73 rows=173 width=20) (actual time=0.001..0.028 rows=173 loops=1)
                                         ->  Index Only Scan using resource_env_model_agent_index on resource r  (cost=0.42..19.36 rows=786 width=48) (actual time=0.002..0.087 rows=2085 loops=173)
                                               Index Cond: ((environment = c.environment) AND (model = c.version))
                                               Heap Fetches: 0
 Planning Time: 2.399 ms
 Execution Time: 61.304 ms
(27 rows)
```
Then it becomes very clear that the last sub-query is very expensive. Also note that every single sub-query is doing a sequential table scan. We really have to improve the index usage. A few ideas to make it perform better:

1) Move the `is_primary` check outside of the NOT IN statement. That way we can trim down the amount of potential agents quickly without performing the expensive NOT IN check.

```
DELETE FROM public.agent
id_primary IS NULL AND ...
```

2) The sub-query that checks whether an agent is still referenced by a resource, performs a join between configurationmodel and resource. I don't think that join is required. It's the resource that requires the agent and not the model. So we could do something like this:

```
DELETE FROM public.agent AS a
WHERE NOT EXISTS(SELECT 1 FROM resource AS re where a.environment=re.environment AND a.name=re.agent)
```
If we do that, we need to make sure that the resource table has an index on (environment, agent).

Reply:
I am going to create a ticket to make it easier to do performance tests of queries, because the current approach is not very developer friendly.

Reply:
I think there is a bug in these sub-queries. They don't filter on the environment that is used by the surrounding Agent record. Is that correct?

Reply:
I'm not sure I understand the bug you are referring to, but I've updated the query so maybe I've solved it?

Reply:
@arnaudsjs regarding the index there is already this one:
`CREATE INDEX resource_env_model_agent_index ON resource (environment, model, agent);`
Do you think it is worth it to still add another one without the model?

Reply:
An index on `(environment, model, agent)` cannot be used to do a lookup for a `(environment, agent)` pair, is it? You can only use that index to filter on `environment`, but that column is already part of the primary key.